coreboot-toolchain: Update runtime dependencies

felixsinger opened the pull request:
###### Motivation for this change
Add the build option `installExtraPackages`, which allows switching
between a "basic" and an "extended" toolchain. This option is set to
true by default adding additional packages to propagatedBuildInputs.
These packages are essential to make the toolchain fully functional.
Most developers and users will need them. For some usecases, these
packages might not be needed. So this option can be set to false.

This includes:

  * git
  * ncurses
  * pkg-config
  * python3Minimal

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

felixsinger committed:
coreboot-toolchain: Update runtime dependencies

Add the build option `installExtraPackages`, which allows switching
between a "basic" and an "extended" toolchain. This option is set to
true by default adding additional packages to propagatedBuildInputs.
These packages are essential to make the toolchain fully functional.
Most developers and users will need them. For some usecases, these
packages might not be needed. So this option can be set to false.

This includes:

  * git
  * ncurses
  * pkg-config
  * python3Minimal

Signed-off-by: Felix Singer <felixsinger@posteo.net>

piegamesde commented:
I don't really understand why those dependencies are added to `propagatedBuildInputs` (or what `propagatedBuildInputs` even means for compiled applications).

Furthermore, `python2` is *way* beyond its end of life and we generally want to get rid of it. Are the board configurations depending on it really that important? If so, please at least hide it behind a feature flag.

symphorien reviewed (changes requested):
propagatedBuildInputs is only for libs that coreboot needs to link successfully into a final executable. I don't think this is the case here. Also this probably breaks cross as git is in buildInputs and not in nativeBuildInputs. Maybe propagatedNativeBuildInputs could be used instead, but honestly I'm not sold on the overall idea. Many libs are meant to be used with pkg-config but I have never seen one propagate it.

felixsinger commented:
> I don't really understand why those dependencies are added to `propagatedBuildInputs` (or what `propagatedBuildInputs` even means for compiled applications).

`propagatedBuildInputs` is documented here https://nixos.org/guides/nix-pills/basic-dependencies-and-hooks.html#idm140737319497040.

Basically, both `buildInputs` and `propagatedBuildInputs` are runtime dependencies. The difference is that packages in `buildInputs` are directly related to the package itself, respectively what the package ships. So this is used for libraries against which the software (what the package contains) is linked to, but these are not exposed to the user environment (e.g. through the PATH environment variable).

`propagatedBuildInputs` is used for indirect dependencies. These are not directly related to the package itself, since the software in that package works without them, but they are needed in the environment in which the package is used, which means their PATHs (like library or bin paths) are exposed to the user environment.

> Furthermore, python2 is way beyond its end of life and we generally want to get rid of it. Are the board configurations depending on it really that important?

Yes, it is important. Else I wouldn't add it ;)

It is required for a script which is used at compile time for Intel binary blobs. These are essential for the hardware initialization and all modern platforms depend on them. I pinged an Intel guy to update it, but that might take a while.

>  If so, please at least hide it behind a feature flag.

I can't relate what you mean. Please explain that further.

> Also this probably breaks cross as git is in buildInputs and not in nativeBuildInputs.

Uh, what?

c0d3z3r0 commented:
> I don't really understand why those dependencies are added to `propagatedBuildInputs` (or what `propagatedBuildInputs` even means for compiled applications).
> 
> Furthermore, `python2` is _way_ beyond its end of life and we generally want to get rid of it. Are the board configurations depending on it really that important? If so, please at least hide it behind a feature flag.

Will be fixed soon: https://review.coreboot.org/c/coreboot/+/58088

felixsinger commented:
So, we just found out that the script has been secretly ported to Python 3 a while ago. Just the build system wasn't updated. Removed Python 2 and added Python 3.

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/default.nix:
```diff
@@ -1,6 +1,7 @@
 { lib, stdenvNoCC, fetchurl, fetchgit,
   gnumake, patch, zlib, git, bison,
```
```suggestion
{ lib, stdenvNoCC, fetchurl, fetchgit
, gnumake, patch, zlib, git, bison
```

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/default.nix:
```diff
@@ -1,6 +1,7 @@
 { lib, stdenvNoCC, fetchurl, fetchgit,
   gnumake, patch, zlib, git, bison,
-  flex, gnat11, curl, perl
+  flex, gnat11, curl, perl, ncurses,
+  pkg-config, python3Minimal
```
```suggestion
, flex, gnat11, curl, perl, ncurses
, pkg-config, python3Minimal
```

SuperSandro2000 commented on pkgs/development/tools/misc/coreboot-toolchain/default.nix:
```diff
@@ -66,9 +67,11 @@ in stdenvNoCC.mkDerivation rec {
-  buildInputs = [ gnat11 flex zlib ];
+  buildInputs = [ flex gnat11 zlib ];
+
+  propagatedBuildInputs = [ git ncurses pkg-config python3Minimal ];
```
Are those packages really strictly required at all time for this package to be useful? Can we maybe use makeWrapper?

piegamesde replied:
> Can we maybe use makeWrapper?

No, because as far as I can tell, these are not run-time dependencies of the code packaged here, but they are meant to be further tools that belong into the Coreboot toolchain.

SuperSandro2000 replied:
We don't propagate tools that are nice to have. This is only done if those packages are strictly required and the package is not useable without them.

felixsinger replied:
These are not nice to have. These are required so that the toolchain is fully functional. We can not use makeWrapper since we would have to wrap scripts from the coreboot build system, which is outside of this package.

felixsinger replied:
I just added libuuid since this is required to build TianoCore.

petabyteboy commented:
I would like to have two versions of the toolchain, something like toolchain-minimal and toolchain-full.

The toolchain is functional without any of these deps for some purposes.

For example, I have updated my [corenix](https://git.petabyte.dev/petabyteboy/corenix) project to use the coreboot-toolchain package, and it doesn't need any additional buildInputs or nativeBuildInputs, since it builds only the configured primary and secondary payloads in seperate derivations, some of them without using the coreboot toolchain at all.

felixsinger commented:
> I would like to have two versions of the toolchain, something like toolchain-minimal and toolchain-full.

Sounds like a good compromise. Thanks!

I also planned to add other target architectures, but now I think I will integrate your update script first before I do any other changes. Else, the package file might get a bit messy. https://github.com/felixsinger/nixpkgs/commits/pkgs/cbtoolchain-add-archs

Which naming scheme do you prefer? I intended to use `coreboot-toolchain.<arch>`. What do you think about `coreboot-toolchain.<variant>.<arch>`? So it would look like this `coreboot-toolchain.minimal.i386`.

felixsinger commented:
Seems like lzma causes some issues on Darwin. How can I fix this? https://pipelines.actions.githubusercontent.com/5egjczZns7sw5r3XAkdhglSC3M649O3PU6vvyY1jOtibky1YRq/_apis/pipelines/1/runs/468802/signedlogcontent/3?urlExpires=2021-10-04T00%3A13%3A34.1105711Z&urlSigningMethod=HMACV1&urlSignature=CR64UlVKm1nnSFOi7mG0vgby5THyDvBhWZd6jrQOi08%3D

felixsinger commented:
> According to the CI log, it seems like lzma causes some issues on Darwin. How can I fix this?

Seems like replacing lzma with xz solved the issue.

felixsinger commented:
I had a discussion about this patch with some people today. It seems we all have a different understanding of what is the toolchain and what should be part of it and what should not. Most people will need these four packages, some people might not need them and others told me to add more packages. Many people have many different opinions about this.

However, I feel like people (including myself) are putting too much emotions into this and start overengineering the package, which adds just more maintenance overhead actually. Nobody likes maintenance overhead :)

I decided to close this pull request since this package should fulfill the usecase of everyone (hopefully). Developers who need these packages can just add them to their Nix shell, since they need one for the other packages anyway. So why bother about these four packages?
